### PR TITLE
Tool types

### PR DESCRIPTION
**Pull request by LysandreJik**
WIP PR to have specific types for tool outputs, which should clarify interaction to and from Agents.

Left to do:
- [x] Remove or complete the video integration
- [x] Add support for remote tools
- [x] Complete documentation
- [x] Test it out with real world use-cases
- [x] Add a test to ensure that the inputs are cast correctly (so far only the outputs are tested)
- [x] Arrange the dependencies so that they don't make all the tests fail

### CODE REVIEW

File: src/transformers/tools/agent_types.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
+import inspect
```
Comment (sgugger) [inline src/transformers/tools/agent_types.py:1]: Nit: missing copyright

File: src/transformers/tools/agent_types.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        else:
+            raise ValueError(f"Unsupported type for {self.__class__.__name__}: {type(value)}")
+
+    def _ipython_display_(self, include=None, exclude=None):
+        """
+        Displays correctly this type in an ipython notebook (ipython, colab, jupyter, ...)
+        """
+        from IPython.display import Image, display
+
+        display(Image(self.to_string()))
```
Comment (sgugger) [inline src/transformers/tools/agent_types.py:97]: You can load an image from a string (so filename?) like this? Isn't an `Image.open` required?

File: src/transformers/tools/agent_types.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    INSTANCE_TYPE_MAPPING[PIL.Image] = AgentImage
+
+
+def handle_agent_inputs(*args, **kwargs):
+    args = [(arg.to_raw() if isinstance(arg, AgentType) else arg) for arg in args]
+    kwargs = {k: (v.to_raw() if isinstance(v, AgentType) else v) for k, v in kwargs.items()}
+    return args, kwargs
+
+
+def old_handle_agent_inputs(*args, tool_class, **kwargs):
```
Comment (sgugger) [inline src/transformers/tools/agent_types.py:194]: Doesn't seem to be used?

**Review by sgugger — APPROVED** [review]
Looks super nice, thanks a lot for adding this!

File: src/transformers/tools/agent_types.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/tools/agent_types.py:111]: The `Image` coming from `IPython.display` requires either a string (path) or bytes. I initially had it set up with bytes, but it ended up being a significant amount of code for little upside, so I chose to go with the string instead,

**Review by sgugger — APPROVED** [review]
Super nice!

**Pull request merged**